### PR TITLE
Fix optional injection bugs

### PR DESCRIPTION
**Pull request by BCSol**
Add tests and fixes for:
1. OptionalActiveDescriptor fails to inject @Named optional values, because (name) qualifiers are not propagated for the retrieval of the wrapped value. As a consequence any bound instance (unnamed, with same name, or other name) may be injected.
This is a treacherous bug, in particular when injecting configuration parameters and migrating from previous versions.
2. Trying to inject Optional<Optional<String>> one might expect the following:
    - Binding a literal string results to Optional<Optional<String>>, but actaally Optional<Optional.empty> is injected
    - Binding Optional<Optional<String> results to Optional<Optional<String>, but actually Optional<Optional<Optional<String>>> is injected    
    Furthermore, one might expect that more specific bindings take precedence over less specific bindings, that is also not the case.
    It is true that nested Optionals are not a common use case, but the precedence should be implemented either way.

3. Trying to inject Guice services as Optional into hk2 services always causes the optional to be empty.

### CODE REVIEW

File: hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/optional/NamedOptionalInjection.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (c) 2012, 2018 Oracle and/or its affiliates. All rights reserved.
```
Comment (smillidge) [inline hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/optional/NamedOptionalInjection.java:2]: Copyright shouldn't be Oracle on a new file you can either use;

`Copyright (c) {year} Contributors to the Eclipse Foundation`

or 

`Copyright (c) {year} {owner}`

See https://www.eclipse.org/projects/handbook/#ip-copyright-headers

Also year should be 2020

Comment (BCSol) [reply hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/optional/NamedOptionalInjection.java]: Hey sorry for the inconvenience :)
I changed the licence according to the link.

**Comment by BCSol** [comment]
I found some more related bugs, which I wrote test for and bundled them here together.

**Comment by smillidge** [comment]
This is failing Jenkins build

**Comment by BCSol** [comment]
> This is failing Jenkins build

tbh, I am a bit confused as to why.
```
19:01:09  [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:testCompile (default-testCompile) on project hk2-locator: Compilation failure: Compilation failure: 
19:01:09  [ERROR] /home/jenkins/agent/workspace/hk2_PR-516/hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/optional/NamedOptionalTest.java:[91,43] cannot find symbol
19:01:09  [ERROR]   symbol:   method isEmpty()
19:01:09  [ERROR]   location: variable providedOptional of type java.util.Optional<java.lang.String>
19:01:09  [ERROR] /home/jenkins/agent/workspace/hk2_PR-516/hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/optional/NamedOptionalTest.java:[106,43] cannot find symbol
19:01:09  [ERROR]   symbol:   method isEmpty()
19:01:09  [ERROR]   location: variable providedOptional of type java.util.Optional<java.lang.String>
19:01:09  [ERROR] /home/jenkins/agent/workspace/hk2_PR-516/hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/optional/NestedOptionalTest.java:[66,53] cannot find symbol
19:01:09  [ERROR]   symbol:   method isEmpty()
19:01:09  [ERROR]   location: class java.util.Optional<java.lang.String>
19:01:09  [ERROR] /home/jenkins/agent/workspace/hk2_PR-516/hk2-locator/src/test/java/org/glassfish/hk2/tests/locator/optional/NestedOptionalTest.java:[102,53] cannot find symbol
19:01:09  [ERROR]   symbol:   method isEmpty()
19:01:09  [ERROR]   location: class java.util.Optional<java.lang.String>
19:01:09  [ERROR] -> [Help 1]
19:01:09  [ERROR] 
19:01:09  [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
19:01:09  [ERROR] Re-run Maven using the -X switch to enable full debug logging.
19:01:09  [ERROR] 
19:01:09  [ERROR] For more information about the errors and possible solutions, please read the following articles:
```
The error log seems to indicate, that `java.util.Optional` doesn't have the method `isEmpty()`, which doesn't make much sense.
@smillidge could you rerun the pipeline or do you have an idea what could be causing this?

**Comment by smillidge** [comment]
IsEmpty is JDK 11 only and we need to support JDK 8 fir Jakarta EE 9

**Review by smillidge — APPROVED** [review]

**Pull request merged**